File: leetcode/practice/13-roman-to-integer/hash.cpp
class Solution
{
    unordered_map<char , int> hash =
    {
        { 'I' , 1 },
        { 'V' , 5 },
        { 'X' , 10 },
        { 'L' , 50 },
        { 'C' , 100 },
        { 'D' , 500 },
        { 'M' , 1000 }
    } ;

public:
    int romanToInt(string s)
    {
        int len = s.length() ;
        int num = 0 ;
        for(int i = len - 1 ; i >= 0 ; i--)
        {
            if(i < len - 1 and hash[s[i]] < hash[s[i + 1]])
                num -= hash[s[i]] ;
            else 
                num += hash[s[i]] ;
        }
        return num ;
    }
};

File: leetcode/practice/13-roman-to-integer/hash.java
class Solution
{
    public int romanToInt(String s)
    {
        HashMap<Character , Integer> roman = new HashMap<>() ;
        roman.put('I' , 1) ;
        roman.put('V' , 5) ;
        roman.put('X' , 10) ;
        roman.put('L' , 50) ;
        roman.put('C' , 100) ;
        roman.put('D' , 500) ;
        roman.put('M' , 1000) ;
        int ans = 0 ;
        for(int i = s.length() - 1 ; i >= 0 ; i--)
        {
            if(i < s.length() - 1 && roman.get(s.charAt(i)) < roman.get(s.charAt(i + 1)))
                    ans -= roman.get(s.charAt(i)) ;
            else
                ans += roman.get(s.charAt(i)) ;
        }
        return ans ;
    }
}
